classes/task/collect_metrics_task.php

```diff
@@ -0,0 +1,163 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_cloudmetrics\task;
+
+use tool_cloudmetrics\metric;
+use tool_cloudmetrics\collector;
+
+/**
+ * Controls the running of metrics within a single task.
+ *
+ * Metrics are measured at specific time determined by their frequency. For example, a metric set to FREQ_30MIN
+ * is supposed to be measured every thirty minutes.
+ *
+ * The metrics are required to measured in sync with each other. For example, all FREQ_30MIN are required to
+ * be measured at the same time, and those of FREQ_15MIN will also be measured at this time. And so on.
+ *
+ * This is done by taking a suitable reference time, and measuring the number of minutes that have passed since
+ * then, and determining if a whole number of intervals have passed since then.
+ *
+ * For example if 75 minutes have passed since 'midnight last Sunday', then FREQ_15MIN metrics will be measured,
+ * but FREQ_30MIN metrics will not.
+ *
+ * @package   tool_cloudmetrics
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+class collect_metrics_task extends \core\task\scheduled_task {
+
+    const FIVEMINUTES = 5;
+    const FIFTEENMINUTES = 15;
+    const THIRTYMINUTES = 30;
+    const ONEHOUR = 60;
+    const THREEHOURS = 180;
+    const TWELVEHOURS = 720;
+    const ONEDAY = 1440;
+    const ONEWEEK = 10080;
+
+    private $time = null;
+
+    /**
+     * Display name for the task name.
+     *
+     * @return \lang_string|string
+     * @throws \coding_exception
+     */
+    public function get_name() {
+        return get_string('task_runner', 'tool_cloudmetrics');
+    }
+
+    /**
+     * Set a specific time for the use. Defaults to now.
+     *
+     * @param int $time
+     */
+    public function set_time(int $time) {
+        $this->time = $time;
+    }
+
+    /**
+     * Will send the metrics to the collectors.
+     *
+     * @param array $items
+     */
+    public function send_metrics(array $items) {
+        collector\manager::send_metrics($items);
+    }
+
+    /**
+     * Determine the highest level of frequency to be used.
+     *
+     * Returns the highest frequency for which a whole number of time intervals fits in the value given.
+     *
+     * Each interval (except monthly) is a whole number of the next smallest interval.
+     * So if (for example), it is determined that all 3 hour metrics are to be measured, all
+     * metrics for smaller intervals will also be measured.
+     *
+     * @param int $timediff The number of minutes since the reference time.
+     * @return int
+     */
+    public function get_frequency_cutoff(int $timediff): int {
+        if ($timediff % self::ONEWEEK == 0) {
+            return metric\manager::FREQ_WEEK;
```

Comment:
I think this would read a little cleaner as a switch. Less brackets control statements. idk, try it out and see what it reads like

Reply:
It needs to be done in a specific order. So unless PHP guarantees the order of comparisons, it needs to be done with if ... else if.

Reply:
Right, I see now, as divisions of 1 day will also be divisions of 12 hours. I think its worth commenting this, and also maybe extracting to a new function. My gut still tells me there is a more clear way, building UP from the smallest increment perhaps. especially because if you cant divide 5 minutes in equally, you can never divide 1 hour in equally. 

Doesnt look like this code is performance critical, so might not be worth the effort, but an  explanation comment for above would be great

Reply:
I have reversed the test order. It is already in a function of it's own, and it describes the matryoshka nature of the intervals in the function docs.